server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java

```diff
@@ -262,4 +348,66 @@ static BytesRef toBytesRef(BytesRef scratch, String v) {
         scratch.length = UnicodeUtil.UTF16toUTF8(v, 0, v.length(), scratch.bytes);
         return scratch;
     }
+
+    /**
+     * Build a {@link LeafIteratorLookup} which checks for norms of a text field.
+     */
+    public static LeafIteratorLookup lookupMatchingAll() {
+        return new LeafIteratorLookup() {
+            @Override
+            public DocIdSetIterator lookup(LeafReaderContext ctx) throws IOException {
+                return DocIdSetIterator.all(ctx.reader().maxDoc());
+            }
+
+            @Override
+            public String toString() {
+                return "All";
+            }
+        };
+    }
+
+    /**
+     * Build a {@link LeafIteratorLookup} which checks for the field in the
+     * {@link FieldNamesFieldMapper field names field}.
+     */
+    public static LeafIteratorLookup lookupFromFieldNames(String fieldName) {
+        return new LeafIteratorLookup() {
+            private final BytesRef name = new BytesRef(fieldName);
+
+            @Override
+            public DocIdSetIterator lookup(LeafReaderContext ctx) throws IOException {
+                Terms terms = ctx.reader().terms(FieldNamesFieldMapper.NAME);
+                if (terms == null) {
+                    return null;
```

Comment:
Should we return `DocIdSetIterator.all(ctx.reader().maxDoc())` when the terms don't exist? I think we allow disabling the `_field_names` field on 7.x. However, we would lose the benefit for fields without values in a segment. We can check the created index version, but we don't have it here.

Reply:
Oh boy! That's fun. Let me play a bit.